var/spack/repos/builtin/packages/mesa/package.py

```diff
@@ -180,7 +180,8 @@ def configure_args(self):
     @property
     def libs(self):
         for dir in ['lib64', 'lib']:
-            libs = find_libraries('libGL', join_path(self.prefix, dir),
+            libs = find_libraries(['libGL', 'libOSMesa'],
+                                  join_path(self.prefix, dir),
                                   shared=True, recursive=False)
             if libs:
```

Comment:
@adamjstewart @v-dobrev is it intentional to return `None` here instead of an empty list? This seams to be generally problematic when setting up the build environment, see the PR description error message.

Reply:
Returning `None` informs the `libs` handler that the search failed (and reporting it as an error is the right behavior). Simply returning an empty list is not treated as an error, see https://github.com/spack/spack/blob/4cd537f6c7153675fde2ce0bb66a2fb9e9fc0a9a/lib/spack/spack/spec.py#L862-L891

Note that the default handler (used if the `libs` property is not explicitly defined by a package) works in a similar manner: https://github.com/spack/spack/blob/4cd537f6c7153675fde2ce0bb66a2fb9e9fc0a9a/lib/spack/spack/spec.py#L792-L799

The raised exception is different but that is just unintentional inconsistency.

Reply:
Thanks for all the details :+1:

Reply:
Thanks for all the details :+1: 
Why do we set the property at all, isn't the default implementation searching for all shared libs sufficient?

Reply:
The default search uses the name of the package only, so for `mesa` it will search for `libmesa.{a,so,dylib}` which is not the real name of the library, see the code just above the second link I posted above.

In `glvis`, we use `spec['gl'].libs` which is then delegated by spack to the specific provider for the virtual `gl` package -- the default for that being `mesa`.

Reply:
ohhh, thanks again. TIL!